src/platform/i386/exception.c

```diff
@@ -13,19 +13,47 @@ print_regs_state(struct pt_regs *regs)
 {
 	PRINTK("registers:\n");
 	PRINTK("General registers-> EAX: %x, EBX: %x, ECX: %x, EDX: %x\n", regs->ax, regs->bx, regs->cx, regs->dx);
-	PRINTK("Segment registers-> CS: %x, DS: %x, ES: %x, FS: %x, GS: %x, SS: %x\n", regs->cs, regs->ds, regs->es,
-	       regs->fs, regs->gs, regs->ss);
-	PRINTK("Index registers-> ESI: %x, EDI: %x, EIP: %x, ESP: %x, EBP: %x\n", regs->si, regs->di, regs->ip,
-	       regs->sp, regs->bp);
+	PRINTK("Segment registers-> CS: %x, DS: %x, ES: %x, FS: %x, GS: %x, SS: %x, ESI: %x, EDI: %x \n", regs->cs, regs->ds, regs->es,
+	       regs->fs, regs->gs, regs->ss, regs->si, regs->di);
+	PRINTK("Index registers-> EIP: %x, ESP: %x, EBP: %x\n", regs->ip, regs->sp, regs->bp);
 	PRINTK("Indicator-> EFLAGS: %x\n", regs->flags);
 	PRINTK("(Exception Error Code-> ORIG_AX: %x)\n", regs->orig_ax);
 }
 
+void
+fault_regs_save(struct pt_regs *regs, struct thread *thd)
+{
+	copy_all_regs(regs, &(thd->fault_regs));
+}
+
+void
+fault_handler_sinv(struct pt_regs *regs, capid_t cap)
+{
+	struct cos_cpu_local_info *ci       = cos_cpu_local_info();
+	struct thread *            curr_thd = thd_current(ci);
+	struct cap_header *        fh;
+	struct comp_info *         cos_info;
+	thdid_t                    thdid = curr_thd->tid;
+	unsigned long              ip, sp;
+	u32_t                      fault_addr = 0, errcode, eip;
+
```

Comment:
I think before you `SINV` into the fault-handler, we should first see if it is a kernel-level fault or user-level fault. I don't think you'd succeed, perhaps leads to double-fault if you tried anything intelligent if it is a kernel-level fault (obviously it depends on the type of fault.) but I think there is no way a user-level handler would handle kernel-level faults (bugs) obviously. Right?

Reply:
I might be wrong here about "obviously it depends on the type of fault". 
I need to read myself but will the user-level be able to resolve any kind of kernel-level fault, even just a page-fault? I doubt it! 
In my opinion, all kernel level faults are our kernel `BUGS` in someway or the other.